components/places/src/bookmark_sync/incoming.rs

```diff
@@ -33,30 +34,42 @@ impl<'a> IncomingApplicator<'a> {
         Self { db }
     }
 
-    pub fn apply_payload(
-        &self,
-        payload: sync15::Payload,
-        timestamp: ServerTimestamp,
-    ) -> Result<()> {
-        if payload.is_tombstone() {
-            self.store_incoming_tombstone(
-                timestamp,
-                BookmarkRecordId::from_payload_id(payload.id).as_guid(),
-            )?;
-        } else {
-            let value: JsonValue = payload.into();
-            match value["type"].as_str() {
-                Some("bookmark") => self.store_incoming_bookmark(timestamp, &value)?,
-                Some("query") => self.store_incoming_query(timestamp, &value)?,
-                Some("folder") => self.store_incoming_folder(timestamp, &value)?,
-                Some("livemark") => self.store_incoming_livemark(timestamp, &value)?,
-                Some("separator") => self.store_incoming_sep(timestamp, &value)?,
-                _ => {
-                    return Err(
-                        ErrorKind::UnsupportedIncomingBookmarkType(value["type"].clone()).into(),
-                    )
-                }
-            };
+    pub fn apply_payload(&self, record: IncomingBso) -> Result<()> {
+        // This is a little odd - we need to go via serde_json::Value so we can
+        // work out what we want to deserialize it as.
+        let timestamp = record.envelope.modified;
+        let json_content = record.to_content::<serde_json::Value>();
+        match json_content.kind {
+            IncomingKind::Tombstone => {
+                self.store_incoming_tombstone(
+                    timestamp,
+                    BookmarkRecordId::from_payload_id(json_content.envelope.id.clone()).as_guid(),
+                )?;
+            }
+            IncomingKind::Content(value) => {
+                match value["type"].as_str() {
+                    Some("bookmark") => self.store_incoming_bookmark(timestamp, &value)?,
+                    Some("query") => self.store_incoming_query(timestamp, &value)?,
+                    Some("folder") => self.store_incoming_folder(timestamp, &value)?,
+                    Some("livemark") => self.store_incoming_livemark(timestamp, &value)?,
+                    Some("separator") => self.store_incoming_sep(timestamp, &value)?,
+                    _ => {
+                        // XXX - we probably should treat this like malformed so the entire
+                        // sync doesn't fail?
+                        // (Either that, or have ::Malformed below fail too?)
```

Comment:
I personally like using `support::error::report_error!` for both of them, then continuing on.  That way we escalate it to a sentry report, but still try to continue syncing for the user.